### DISCUSSION TITLE
간단한 유틸 함수 NPM 라이브러리 배포해보기 (feat. TypeScript 지원, ESM 지원) - 정현수 기술 블로그

### DISCUSSION THREAD

**Discussion by giscus** [junghyeonsu/junghyeonsu.dev.comment#44] [Announcements]
# 간단한 유틸 함수 NPM 라이브러리 배포해보기 (feat. TypeScript 지원, ESM 지원) - 정현수 기술 블로그

NPM 배포를 처음 해보는 사람들을 위해서 차근차근 가이드를 작성해봤습니다.

https://junghyeonsu.com/posts/deploy-simple-util-npm-library/

**Comment by developer-1px** [comment]
와!! 정말 상세하고 쉽게 핵심을 다 적어주셨네요!! npm배포가 한번 하고 나면 금방 익숙해지는데 모르는 상태로 한번 해보려고 하면 그 과정까지 만나는 에러들로 하여금 엄청 지치고 힘들어서 무섭게 느껴지는데 좋은 글 너무너무 고맙습니다!!

**Reply by junghyeonsu** [reply]
읽어주셔서 감사해요 테오~

**Comment by yogjin** [comment]
최근 ts, esm, cjs를 모두 지원하는 라이브러리 배포에 관심이 생겼는데,
딱 맞는 글을 올려주셔서 참고할 수 있겠네요. 감사합니다!

**Reply by junghyeonsu** [reply]
읽어주셔서 감사합니다

**Comment by joohaem** [comment]
글이 너무 깔끔해요!! 정리된 내용 감사합니다-

**Reply by junghyeonsu** [reply]
읽어주셔서 감사합니다ㅎㅎ

**Comment by bae-sh** [comment]
덕분에 npm 배포 뿐 아니라 cjs,esm 모두 지원하는 듀얼 패키지를 만들 수 있었습니다! 감사합니다!!

**Comment by minmoong** [comment]
글 잘 읽었습니다! 질문 하나만 해도 될까요?

node-fetch 라이브러리를 제가 만들 패키지에서 import 하는데, 해당 라이브러리가 cjs를 지원하지 않아서(esm 라이브러리라서) 타입스크립트를 cjs로 컴파일 한 뒤에 빌드된 결과물을 실행하면 esm 라이브러리인 node-fetch를 require 할 수 없다는 오류가 발생하네요.

이런 경우에는 node-fetch 때문에 cjs로 컴파일을 못하니, 제가 만들 패키지도 cjs를 지원하지 못하는 것이 맞죠?
cjs를 지원하는 다른 http 요청 라이브러리를 찾아야 하는건가요?

**Reply by junghyeonsu** [reply]
안녕하세요 minmoong님 글 읽어주셔서 감사합니다.
제가 살펴보니 https://www.npmjs.com/package/node-fetch-commonjs 라이브러리가 따로 있는 것 같은데
해당 라이브러리가 나온 이유가 minmoong님이 고민하시는 이유와 비슷한 이유에서 나온 것 같아요ㅎㅎ 한 번 살펴보시면 좋을 것 같네요.

그리고 node 16버전의 지원이 끊키고, 18버전에서 fetch API를 쓸 수 있는 걸로 알고있어서
node-fetch와 같은 라이브러리가 필요한지도 생각을 해봐야겠네요!
다양한 환경에서 지원하고자 한다면 node-fetch 같은 라이브러리를 사용해서 지원해주는 방향도 있겠네요
이건 minmoong님이 어떤 라이브러리, 어떤 타겟의 유저를 생각하고 계신지에 따라 달라지겠네요.

**Reply by junghyeonsu** [reply]
그리고 기본적으로 esm에서 cjs를 사용하는건 되지만, cjs 패키지에서 esm 패키지를 사용할 수는 없어요

**Comment by zaeyonn** [comment]
감사합니다~

**Comment by moolmin** [comment]
npm 배포의 전반적인 흐름을 이해했어요! 좋은글 감사드립니다 :)

**Comment by tobesimple7** [comment]
npm 배포를 한번도 안해본 탓에 잘 몰랐는데요.
이렇게 자세히 안내 해주셔서 고마워요.
덕분에 많은 도움이 되었습니다.

**Comment by bohodays** [comment]
npm 배포에 관심이 생겨서 읽게되었는데, 가독성이 좋아서 이해가 잘 되었습니다.
좋은 글 감사합니다 !
많은 도움이 되었습니다 :)
